### ISSUE TITLE
GTK vs QT

### ISSUE THREAD

**Issue by ben2talk** [Qalculate/qalculate-qt#11]
Doing some maths homework here, there are some issues.

#1 With the GTK history, I can select to REMOVE only one history item.
With QT, I can only select all, or clear (all).

We need the options available in GTK.

#2 with GTK preferences>Numbers & Operators we can select Division and Multiplication signs.
Under QT this is absent.

**Comment by hanna-kn** [comment]
I will add an option to use dot operator ("⋅") as multiplication sign. The main reason the dot operator is not used by default is because of the better visibility of "×". It is however always used for output of unit expressions.

I'm not sure if it is a good idea to add the option to use ÷ as default division sign, and the division slash Unicode symbol ("∕") is often typographically problematic for user input (it is however used for output).

Handling of single history items is a bit trickier in the Qt version because of the use of a text widget instead of a list.

The initial Qt version does not include all planned features.

**Comment by Ygarr** [comment]
FLTK :)

**Comment by funnym0nk3y** [comment]
Is the Qt GUI intended to reach feature parity with the GTK GUI eventually?

**Comment by hanna-kn** [comment]
> Is the Qt GUI intended to reach feature parity with the GTK GUI eventually?

Yes. Although the functionality might be implemented differently. The main features currently missing (from my perspective) is configure keyboard shortcuts and customizable keypad. Is there any specific feature you're missing?

**Comment by funnym0nk3y** [comment]
I haven't used the Qt version much, as I am way more comfortable with the GTK one. But I will try to comment on a few things: 

1. The keypad in the GTK version offers many more functions which makes usage more approachable, as I can see what is available. This is the main reason I dislike fully fledged CAS as it always requires a whole week of RTFM. 
2. The GTK version allows the usage with only the keyboard. But I think that is what you meant by keyboard shortcuts. 
3. There is a small row of text between the input and the result in the GTK version which displays the current settings (The famous calculation in degree instead of radian mistake). 
4. The most recent calculation is better emphasized in the GTK version due to being completely separate from the history. 
5. I think text instead of icons is more comprehensive which is closely related to get going quickly. 
6. I really like that the functions and constants are now in a separate window and not nested deeply in a menu structure. 
7. The Qt GUI starts way faster. 
8. There is a dark mode in Qt. 
9. Qt can be used better on touchscreens. 
10. The Qt version misses the date, vector and matrix insertion helpers. 

Those are the things that caught my eye. Please mind they are very opionated.

**Comment by hanna-kn** [comment]
>     1. The keypad in the GTK version offers many more functions which makes usage more approachable, as I can see what is available. This is the main reason I dislike fully fledged CAS as it always requires a whole week of RTFM.

This is definitely an area that I would like to improve in the Qt GUI. On the other hand I would prefer to steer users towards the completion functionality of the expression entry.
 
>     2. The GTK version allows the usage with only the keyboard. But I think that is what you meant by keyboard shortcuts.

There is no major difference in what can or cannot be done using the keyboard in the two GUI. The GTK version does however provide the possible to add your own keyboard shortcuts.

>     3. There is a small row of text between the input and the result in the GTK version which displays the current settings (The famous calculation in degree instead of radian mistake).

I decided not to add this to avoid clutter, and since the flat mode menu provides quicker access. In most cases the angle unit is displayed in the parsed expression.

>     4. The most recent calculation is better emphasized in the GTK version due to being completely separate from the history.

It was design decision to make the history more prominent. The most recent result is displayed with a bigger font size, but it is of course not comparable to the separate result display in the GTK GUI (which is however cumbersome for long results).

>     5. I think text instead of icons is more comprehensive which is closely related to get going quickly.

Text does however require a lot of space. I do however agree that the GTK version should generally be easier for new users.

>     6. I really like that the functions and constants are now in a separate window and not nested deeply in a menu structure.

The variables, functions, and units windows are also available in the GTK version (using the Edit menu, keypad buttons, or associated shortcuts)

>     10. The Qt version misses the date, vector and matrix insertion helpers.

They are available in the the context menu of the expression entry.

**Comment by montyx99** [comment]
My main issue, there are no keypad handler buttons. Such as I've made it float by accident, and I cannot make it part of the main window again. Also I cannot hide/show the whole keypad.

**Comment by hanna-kn** [comment]
> My main issue, there are no keypad handler buttons. Such as I've made it float by accident, and I cannot make it part of the main window again.

Double-click the title bar of the keypad, or drag the the title bar of the floating keypad to the desired position (or, if not using Windows, press the button left of the x in the title bar keypad).

> Also I cannot hide/show the whole keypad.

Press the keypad button (the rightmost of the left-aligned buttons) and select "None" or desired keypad variant, press the x button (upper right) of the keypad, (un)check "Keypad" in the context menu of the toolbar or the keypad title bar, or press Ctrl+K.

**Comment by gudvinr** [comment]
I detached keyboard but there is no way to attach it back to the window

**Comment by hanna-kn** [comment]
> I detached keyboard but there is no way to attach it back to the window.

See my previous comment (press the button left of the x in the title bar keypad, double-click the title bar of the keypad, or drag the the title bar of the floating keypad to the desired position). I will probably add an option to restore the dock widget defaults in the next version.

> Also for some reason on KDE (which is Qt-native platform) qalculate looks different from other Qt apps. Buttons, selectors, etc have slightly different colors and shapes.

Perhaps Qalculate is using Qt6 (without KDE themes support).

**Comment by yw4z** [comment]
might be off topic but do you planning to release qt / gtk versions seperately. Current portable zip for windows has 150+ mb size . im sure only qt will drop below 50mb. btw qt version works better on windows but one feature is missing to simplify UI. hiding left keypad

**Comment by hanna-kn** [comment]
> might be off topic but do you planning to release qt / gtk versions seperately. Current portable zip for windows has 150+ mb size . im sure only qt will drop below 50mb.

I estimate that the size difference would be less than 20%.

> but one feature is missing to simplify UI. hiding left keypad

In the upcoming version, when the right keypad is hidden, the number pad will be available as a separate keypad (the "Hide Number Pad" option is replaced by "Always Show Number Pad").

**Comment by yw4z** [comment]
Thanks for your reply and effort on this. yes you right too many libraries used as common for qt and gtk. and one more thing, i like design but a bit heavy and 150mb a bit much for my use. i mostly use basic calculation, using history and conversion. any plans for making a minimal version with less features to making it more compact and faster launch. maybe making libraries / modules optional can be solution

**Comment by Jaanyaar** [comment]
> > ```
> > 4. The most recent calculation is better emphasized in the GTK version due to being completely separate from the history.
> > ```
> 
> It was design decision to make the history more prominent. The most recent result is displayed with a bigger font size, but it is of course not comparable to the separate result display in the GTK GUI (which is however cumbersome for long results).
> 
Is there a solution to hide history in QT version? Or a separate window like GTK? Any future plan?

**Comment by hanna-kn** [comment]
> > It was design decision to make the history more prominent. The most recent result is displayed with a bigger font size, but it is of course not comparable to the separate result display in the GTK GUI (which is however cumbersome for long results).
> 
> Is there a solution to hide history in QT version? Or a separate window like GTK? Any future plan?

No.

It might be possible to add options to further increase the prominence of the most recent result (and decrease the prominence of the history), e.g. further size differences, bold font, text colour, background colour, increased distance, and more prominent separator.
You can almost hide the history by reducing the size as much as possible.
